dapps/assets/src/components/contacts/AddContact.ts

```diff
@@ -0,0 +1,130 @@
+  write to the Free Software Foundation, Inc., 51 Franklin Street,
+  Fifth Floor, Boston, MA, 02110-1301 USA, or download the license from
+  the following URL: https://evan.network/license/
+*/
+
+// vue imports
+import Component, { mixins } from 'vue-class-component';
+
+// evan.network imports
+import { EvanComponent, EvanForm } from '@evan.network/ui-vue-core';
+import { ContactsService } from './ContactsService';
+
+@Component
+export default class AddContactComponent extends mixins(EvanComponent) {
+  contactService: ContactsService;
+
+  addressbook;
+
+  idOrEmailValidation = '';
+  idOrEmail = null;
+
+  accountId = null;
+  alias = null;
+  email = null;
+  emailInvite = null;
+  fromAlias = window.localStorage.getItem('evan-alias');
+  msgBody = `${this.$t('_assets.contacts.message-prefill')}${this.fromAlias}`;
+  msgTitle = this.$t('_assets.contacts.subject-prefill');
+
+  async created() {
+    this.initState();
+    const runtime = this.getRuntime();
+    this.contactService = new ContactsService(runtime);
+    this.addressbook = await runtime.profile.getAddressBook();
+  }
+
+  private initState() {
+    this.idOrEmailValidation = '';
+    this.idOrEmail = null;
+
+    this.accountId = null;
+    this.alias = null;
+    this.email = null;
+    this.emailInvite = null;
+    this.fromAlias = window.localStorage.getItem('evan-alias');
+    this.msgBody = `${this.$t('_assets.contacts.message-prefill')}${
+      this.fromAlias
+    }`;
+    this.msgTitle = this.$t('_assets.contacts.subject-prefill');
+  }
+
+  async addContact() {
+    if (this.checkFormValid()) {
+      const formData = {
+        accountId: this.accountId,
+        alias: this.alias,
+        createdAt: new Date().toISOString(),
+        currLang: window.localStorage.getItem('evan-language'),
+        email: this.email,
+        emailInvite: this.emailInvite,
+        fromAlias: this.fromAlias,
+        msgBody: this.msgBody,
+        msgTitle: this.msgTitle,
+        updatedAt: new Date().toISOString()
+      };
+      await this.contactService.addContact(formData);
+
+      this.closePanel();
+
+      this.$emit('contact-added');
+    }
+  }
+
+  handleIdOrEmailChange(value: string) {
+    this.idOrEmailValidation = this.validateIdOrEmail(value);
+    console.log(this.idOrEmailValidation);
+  }
+
+  private validateIdOrEmail(value: string): string {
+    if (EvanForm.validEthAddress(value)) {
+      if (this.addressbook.profile[value]) {
+        return '_assets.contacts.error-added';
+      } else {
+        this.emailInvite = false;
+        this.accountId = value;
+        this.email = null;
+        return '';
+      }
+    } else if (EvanForm.validateEmail(value)) {
+      this.emailInvite = true;
+      this.email = value;
+      this.accountId = null;
+      return '';
+    } else {
+      return '_assets.contacts.error-id-or-email';
+    }
+  }
+
+  checkFormValid(): boolean {
+    if (
+      document.querySelector('#contactForm :invalid') ||
```

Comment:
this looks hacky, I wouldn't rely on browser form validation to be honest.

Reply:
Since it only checks required status, I think this will work very reliably. I'm more confident in native browser checks than custom js validation tbh :)

Reply:
Since it only checks required status, I think this will work very reliably. I'm more confident in native browser checks than custom js validation tbh :)

Reply:
Would also suggest the same.
Hint again: We can also use, the `EvanForm` declaration without automatic formular generation. Just use the `EvanForm` as `value` and validation "holder". Code is not tested, only copied from somewhere :D 

```ts
this.form = (<IssueFormInterface>new EvanForm(this, {
  idOrEmail: {
    value: '',
    validate: function(vueInstance: IssueComponent, form: IssueFormInterface) {
      if (EvanForm.validEthAddress(this.value)) {
        if (vueInstance.addressbook.profile[this.value]) {
          return '_assets.contacts.error-added';
        } else {
          this.emailInvite = false;
          vueInstance.accountId = value;
          vueInstance.email = null;
          return '';
        }
      } else if (EvanForm.validateEmail(this.value)) {
        vueInstance.emailInvite = true;
        vueInstance.email = value;
        vueInstance.accountId = null;
        return '';
      } else {
        return '_assets.contacts.error-id-or-email';
      }
    }
  },
```

```html
<evan-form-control-input
  v-model="form.idOrEmail.value"
  :label="'_assets.contacts.id-or-email' | translate"
  :placeholder="$t('_assets.contacts.id-or-email-placeholder')"
  :required="true"
  :error="form.idOrEmail.error"
  @input="handleIdOrEmailChange"
/>

...

<evan-button
  type="primary"
  native-type="submit"
  class="ml-3 flex-grow-1"
  :disabled="form.invalid"
  :label="'_assets.contacts.add-contact-btn' | translate"
/>
```

Reply:
This would require refactoring the other inputs into evan form too... I tried it and stashed it for now since it requires more refactoring than I thought :D